presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementSingleDistinctAggregation.java

```diff
@@ -0,0 +1,148 @@
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.AggregationNode.Aggregation;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static com.facebook.presto.sql.planner.plan.Patterns.aggregation;
+
+/**
+ * Implements distinct aggregations with similar inputs by transforming plans of the following shape:
+ * <pre>
+ * - Aggregation
+ *        GROUP BY (k)
+ *        F1(DISTINCT s0, s1, ...),
+ *        F2(DISTINCT s0, s1, ...),
+ *     - X
+ * </pre>
+ * into
+ * <pre>
+ * - Aggregation
+ *          GROUP BY (k)
+ *          F1(x)
+ *          F2(x)
+ *      - Aggregation
+ *             GROUP BY (k, s0, s1, ...)
+ *          - X
+ * </pre>
+ * <p>
+ * Assumes s0, s1, ... are symbol references (i.e., complex expressions have been pre-projected)
+ */
+public class ImplementSingleDistinctAggregation
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(ImplementSingleDistinctAggregation::hasDistinct);
+
+    private static boolean hasDistinct(AggregationNode aggregation)
+    {
+        return aggregation.getAggregations()
+                .values().stream()
+                .filter(e -> e.getCall().isDistinct())
+                .map(e -> e.getCall().getArguments())
+                .distinct()
+                .count() == 1;
+    }
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregation, Captures captures, Context context)
+    {
+        Collection<Aggregation> aggregations = aggregation.getAggregations().values();
+
+        List<Set<Expression>> argumentSets = aggregations.stream()
+                .map(Aggregation::getCall)
+                .filter(FunctionCall::isDistinct)
+                .map(FunctionCall::getArguments)
+                .map(HashSet::new)
+                .distinct()
+                .collect(Collectors.toList());
+
+        Set<Set<Expression>> uniqueArgumentSet = ImmutableSet.copyOf(argumentSets);
+
+        if (uniqueArgumentSet.size() != 1 || argumentSets.size() != aggregations.size()) {
```

Comment:
nit: I think you could move this condition to pattern matching and catch `Iterables.getOnlyElement(uniqueArgumentSet)` as capture

Reply:
I don't think `uniqueArgumentSet.size() != 1` is needed at this point as pattern already enforces that. You could do check state though.

Reply:
Will do. I just copied this from the old optimizer

Reply:
> and catch Iterables.getOnlyElement(uniqueArgumentSet) as capture

Are there any examples on how to capture those values in the pattern match? All I see are usages where nodes are captured.